Minor cleanups in core

oscargus opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed
Fixed some minor code issues (duplicate imports, unused variables, etc)

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

oscargus committed:
Minor cleanups in core

oscarbenjamin commented on sympy/core/expr.py:
```diff
@@ -3619,7 +3616,7 @@ def _expand_hint_key(hint):
             if use_hint:
                 hint = '_eval_expand_' + hint
-                expr, hit = Expr._expand_hint(expr, hint, deep=deep, **hints)
+                expr = Expr._expand_hint(expr, hint, deep=deep, **hints)[0]
```
I consider this kind of thing to be defensive programming. If the function is always expected to return a sequence of length 2 then this asserts that it has exactly two items:
```python
>>> a, b = [1, 2, 3]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: too many values to unpack (expected 2)
```
The variable name can also be used to help understand what the function returns even if the `hit` variable is unused.

oscargus replied:
I was also considering this and, e.g., for the mp-function returning unused accuracy variables I did not change it. I see the point to some extent here as well. (And next comment.)

oscarbenjamin commented on sympy/core/mul.py:
```diff
@@ -1113,7 +1113,7 @@ def _matches_noncomm(nodes, targets, repl_dict={}):
         node_ind, target_ind = state
         if node_ind in dictionary:
-            begin, end = dictionary[node_ind]
+            begin = dictionary[node_ind][0]
```
I wouldn't change this.

smichr replied:
But here and above you could use `_` for the unused arg.

oscarbenjamin replied:
I prefer `end`: it shows that the dictionary value is a 2-tuple marking the beginning and end which is helpful information when reading the code.

smichr commented on sympy/core/mul.py:
```diff
@@ -1113,7 +1113,7 @@ def _matches_noncomm(nodes, targets, repl_dict={}):
         node_ind, target_ind = state
         if node_ind in dictionary:
-            begin, end = dictionary[node_ind]
+            begin = dictionary[node_ind][0]
```
```suggestion
            begin, _ = dictionary[node_ind]
```

smichr commented on sympy/core/expr.py:
```diff
@@ -3619,7 +3616,7 @@ def _expand_hint_key(hint):
             if use_hint:
                 hint = '_eval_expand_' + hint
-                expr, hit = Expr._expand_hint(expr, hint, deep=deep, **hints)
+                expr = Expr._expand_hint(expr, hint, deep=deep, **hints)[0]
```
```suggestion
                expr, _ = Expr._expand_hint(expr, hint, deep=deep, **hints)
```

smichr committed:
Apply suggestions from code review

retain defensive programming assertion of the arity of the returned value but use _ to idnicate that it is not needed past that line

oscargus committed:
Fixed comments and removed mutable default arguments

oscargus commented:
Thanks @smichr . I changed them in my local copy (back to using a name, since it may make sense here) and also removed some mutable default arguments as I found #19372 and #20068. Hence, I disabled auto-merge.

oscargus committed:
Fixed merge conflicts

oscargus commented:
There is a bit of `repl_dict.copy()` here which I do not fullt know are required or if they were just a way to avoid mutating the default argument. I've kept them though.

smichr committed:
don't make a second copy of repl_dict

oscargus commented:
@smichr You are right! I had an idea at some stage, but then I automatically replaced all `if ...` with `if ... else .copy` and then I forgot this. Thanks!
